wellcomeml/ml/clustering.py

```diff
@@ -0,0 +1,321 @@
+import umap
+
+from wellcomeml.ml import Vectorizer
+from wellcomeml.logger import logger
+
+
+class TextClustering(object):
+    """
+    Class
+
+    Attributes:
+        vectorizer: The embedding Vectorizer object
+        reducer: A dimensionality reduction object
+        clustering: A clustering model object
+        cluster_ids: Ids of the cluster
+        cluster_names: Names of the clusters
+        cluster_kws: Keywords for the clusters (only if embedding=tf-idf)
+    """
+    def __init__(self, embedding='tf-idf', reducer='umap', clustering='dbscan',
+                 n_kw=10, params={}):
+        """
+        Initialises a cluster pipeline
+
+        Args:
+            embedding(str): 'tf-idf', 'bert', 'keras' or 'doc2vec'.
+            reducer(str): 'umap' or 'tsne'
+            clustering(str): Only 'dbscan' allowed for now
+            n_kw(int): If n_kw > 0 and embedding='tf-idf', the model
+            outputs, besides a cluster id, a list of keyword for each
+            cluster.
+            params(dict): Dictionary containing extra parameters for
+            embedding, reducer or clustering models. Example:
+            {'clustering': {'eps': 0.1}, 'embedding': {'ngram_range': (1,2)}}
+        """
+        self.embedding = embedding
+        self.vectorizer = Vectorizer(embedding=embedding,
+                                     **params.get('embedding', {}))
+        self.n_kw = n_kw
+
+        reducer_dispatcher = {
+            'umap': umap.UMAP,
+            'tsne': TSNE
+        }
+        self.reducer = reducer_dispatcher[reducer](**params.get('reducer', {}))
+
+        clustering_dispatcher = {
+            'dbscan': DBSCAN
+        }
+
+        self.clustering = clustering_dispatcher[clustering](
+            **params.get('clustering', {})
+        )
+
+        self.cluster_ids = None
+        self.cluster_names = None
+        self.cluster_kws = None
+
+    def fit(self, X, *_):
+        """
+        Fits all clusters in the pipeline
+
+        Args:
+            X: A list of texts
+
+        Returns:
+            A TextClustering object
+
+        """
+        self._fit_step(X, step='vectorizer')
+        self._fit_step(step='reducer')
+        self._fit_step(step='clustering')
+
+        if self.embedding == 'tf-idf' and self.n_kw:
+            self._find_keywords(self.embedded_points.toarray(), n_kw=self.n_kw)
+
+        return self
+
+    def _fit_step(self, X=None, step='vectorizer'):
+        """Internal function for partial fitting only a certain step"""
+        if step == 'vectorizer':
+            self.embedded_points = self.vectorizer.fit_transform(X)
+        elif step == 'reducer':
+            self.reduced_points = \
+                self.reducer.fit_transform(self.embedded_points)
+        elif step == 'clustering':
+            self.clustering.fit(self.reduced_points)
+            self.cluster_ids = self.clustering.labels_
+
+    def optimise(self, X, param_grid, n_cluster_range=None, max_noise=0.2,
```

Comment:
is it possible to try pipeline and gridsearch with the memory param to see whether that is redundant?

Reply:
if it is you can still keep the function but possibly utilise the sklearn functions inside to optimise